### ISSUE TITLE
Sample weight support for regression problems?

### ISSUE THREAD

**Issue by kmedved** [pycaret/pycaret#573]
Hello - thanks for this very interesting/promising library. I was wondering whether there is any support or planned support for regression sample weights? These are essential for many types of regression problems, where your observations are not always available at the most granular level and are somehow grouped before being converted to features. This is common with sports data, and with certain types of environmental measurements.

Sklearn and other regression libraries typically implement a `sample_weight` parameter into the `.fit()` call or elsewhere to let the user pass sample weights. Is anything like this possible with PyCaret?

Thanks!

**Comment by pycaret** [comment]
@kmedved You can do that. `create_model` function supports **kwargs argument. Any algorithm that supports sample_weight can be passed directly inside the `create_model` function.

Example:
```
# train lightgbm with specific learning rate
lightgbm = create_model('lightgbm', learning_rate = 0.4)
```

This is just an example to show you can pass any arbitrary parameter inside `create_model` as long as it is supported by the estimator. In this case, `learning_rate` param is supported by `lightgbm`.

Hope this helps.

**Comment by kmedved** [comment]
Thanks @pycaret, but I'm not sure that actually solves this issue. In the lightgbm example, sample weights are passed into a model via the `.fit()` call, not the instantiation call (which is where you'd pass a `learning_rate` parameter for instance).  This isn't specific to LightGBM - other estimators in python take sample weights the same way (e.g., scikit-learn, lightgbm, catboost, xgboost, etc...). They all take sample weights, but none of them take them as a **kwarg - rather they take them with the .fit() call, where you'd pass the X and y datasets.

The reason is that sample weights are not a hyperparameter similar to learning rates. Sample weights are dataset specific, and would presumably need to be passed into pycaret via at the `setup` phase, such as by designating one of the columns of your data the `sample_weight` column. (Similar to how you would pass sample weights into LightGBM at the [dataset construction](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.Dataset.html#lightgbm.Dataset) via the `weight` parameter if using the `.train()` API).

I've set up an example here on Colab showing the issue [here.](https://colab.research.google.com/drive/1CEt2upkKLZyMbU08Nq3GM4DNAxUKFQnR?usp=sharing).

Adding support for weights would be a major improvement to this library and expand the use cases to many new areas. Thanks - I appreciate the help!

**Comment by Yard1** [comment]
I don't think it would be easy to add with how our functions are structured. You can call `fit` directly on the returned model object, though. You can get the training and test data by using `get_config()`. The arguments are (one at a time) `'X_train'`, `'y_train'`, `'X_test'` and `'y_test'`.

**Comment by pycaret** [comment]
@kmedved Thanks a lot for clarifying. I think this is totally possible and I can work on this change fairly quickly. Can you please provide me with 2 working example (one for classification and one for regression) using sklearn directly. If you can share those notebooks I can work on this change.

@Yard1 I think it won't be that hard. We can't use **kwargs but we can have an additional parameter in the `setup` that takes input for `sample_weight` and pass it to .fit() in the codebase. By default, `sample_weight` in `fit` function is set to `None`.

**Comment by pycaret** [comment]
@kmedved I think we can make it work. Here is my working example of how to do it until the functionality is available:

```
from pycaret.datasets import get_data
data = get_data('juice')

from pycaret.classification import *
clf1 = setup(data, target = 'Purchase')

lr = create_model('lr')

# now pass sample weights to lr and fit again
import numpy as np
weights = np.arange(get_config('X_train').shape[0])
lr.fit(X = get_config('X_train'), y = get_config('y_train'), sample_weight = weights)
```

This is working. Once this makes sense to you, we will figure out a way on how to handle this inside pycaret.

**Comment by kmedved** [comment]
Thanks @pycaret - I've put up working examples on [Colab here:](https://colab.research.google.com/drive/1CEt2upkKLZyMbU08Nq3GM4DNAxUKFQnR?usp=sharing.)

Note that if you're passing an evaluation set, into something like LightGBM, xgboost, or Catboost, for the purposes of early stopping, that you need to pass an evaluation set sample sample weight as well. This isn't relevant for sklearn, since they don't do early stopping. I don't know well enough how PyCaret operates internally to know if that's an issue here, but if you're using early stopping, I can add examples to the Colab notebook showing how those validation set sample weights work as well.

I agree that it makes sense to add the `sample_weight `as a parameter in the `setup`, and this would mimic the internal structure of how the various GBDT models handle this internally as well (via the Dataset, Pool, or DMatrix constructors).

**Comment by pycaret** [comment]
@kmedved Actually after writing my comment I discussed this with @Yard1 (he is leading the `2.2` sprint). His view is we should allow `fit_params` within `create_model`, `tune_model`, and all the other trainer functions (except for `compare_models`). Because we have realized that there are few other `params` you can pass using `.fit()` and not all of them are applicable to all the estimators.

Whichever way it would be, I think the end goal here is to allow users to pass any parameter they want to `.fit()` such as `sample_weights`. 

Thanks for providing the example. It will be very helpful for us.

**Comment by kmedved** [comment]
Sure - I'll add examples of each. I presume you want the sklearn API for each? (They each have their own native APIs as well).

One consideration for putting this into `fit_params` vs into `setup` is cross-validation. Insofar as PyCaret is splitting the training data internally via `setup` for cross-validation purposes, you'll need to do the same for the sample weights, so depending how PyCaret handles cross-validation, you may need to have another internal `setup` call regardless when doing cross-validation. Additionally, it seems a bit more intuitive/natural to include the sample weights at the `setup` stage, since that's where all the other data columns (features and labels) are being passed anyway. But I don't understand the inner workings of PyCaret well enough to have a strong view here - just explaining my sense as a user.

**Comment by kmedved** [comment]
[The Colab notebook is updated (sklearn API).](https://colab.research.google.com/drive/1CEt2upkKLZyMbU08Nq3GM4DNAxUKFQnR?usp=sharing.#scrollTo=LI1iTM4zYDlz)

One example of regression and classification for each.

**Comment by pycaret** [comment]
@kmedved Thanks

**Comment by Yard1** [comment]
Yeah now that I think about it that would be the best solution, @pycaret

**Comment by pycaret** [comment]
@kmedved 

This issue has been fixed in the `pycaret-nightly` release. Please install pycaret-nightly using `pip install pycaret-nightly`. Feel free to open a new issue, in case the problem persists. This fix will be published to PyPi as `pycaret==2.2.0` on October 28, 2020.

**Comment by kmedved** [comment]
Great. Thanks @pycaret
